adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/CoverageRDD.scala

```diff
@@ -0,0 +1,148 @@
+
+    val flattenedCoverage = features2Coverage(region)
+
+    if (bpPerBin == 1) {
+      flattenedCoverage // no binning, return raw results
+    } else {
+      // subtract region.start to shift mod to start of ReferenceRegion
+      flattenedCoverage.filter(r => (r.start - region.start) % bpPerBin == 0)
+    }
+  }
+
+  /**
+   * Gets coverage overlapping specified ReferenceRegion. For large, ReferenceRegions,
+   * base pairs per bin (bpPerBin) can be specified to bin together ReferenceRegions of
+   * equal size. The coverage of each bin is the mean coverage over all base pairs in that bin.
+   *
+   * @param region ReferenceRegion to fetch overlapping coverage from
+   * @param bpPerBin base pairs per bin, number of bases to combine to one bin
+   * @return RDD of Coverage Records
+   */
+  def getAggregatedCoverage(region: ReferenceRegion, bpPerBin: Int = 1): RDD[Coverage] = {
+
+    val flattenedCoverage = features2Coverage(region)
+
+    if (bpPerBin == 1) {
+      flattenedCoverage // no binning, return raw results
+    } else {
+      flattenedCoverage.map(r => {
+        // key coverage by binning start site mod bpPerbin
+        // subtract region.start to shift mod to start of ReferenceRegion
+        val key = r.start - ((r.start - region.start) % bpPerBin)
+        (key, r)
+      }).groupByKey
+        .map(r => {
+          // compute average coverage in bin
+          val averageCount = r._2.map(_.count).sum / r._2.map(_.count).size
+          Coverage(r._2.head.contigName, r._1, r._1 + bpPerBin, averageCount)
+        })
+    }
+  }
+
+  /**
+   * Filters FeatureRDD by region and flattens ReferenceRegions to individual base pair
+   * coverage.
+   * @param region Region to filter FeatureRDD
+   * @return Flattened RDD of coverage
+   */
+  private def features2Coverage(region: ReferenceRegion): RDD[Coverage] = {
+    val features = featureRdd.filterByOverlappingRegion(region)
+
+    // flatmap features in region to coverage
+    features.rdd.flatMap(r => {
+      val positions: List[Long] = List.range(r.getStart, r.getEnd)
+      positions.map(n => Coverage(r.getContigName, n, n + 1, r.getScore))
+    }).filter(r => r.start >= region.start && r.start < region.end) // filter out positions from flanking regions
+  }
+
+  /**
+   * Gets flattened RDD of coverage, with coverage at each base pair
+   * @return RDD[Coverage] from underlying FeatureRDD
+   */
+  def flattenedRdd: RDD[Coverage] = {
+    // flatmap features in region to coverage
+    featureRdd.rdd.flatMap(r => {
+      val positions: List[Long] = List.range(r.getStart, r.getEnd)
+      positions.map(n => Coverage(r.getContigName, n, n + 1, r.getScore))
+    })
+  }
+
+  /**
+   * Gets raw RDD of coverage
+   * @return RDD[Coverage] from underlying FeatureRDD
+   */
+  def rdd: RDD[Coverage] = {
+    featureRdd.rdd.map(r => {
+      Coverage(r.getContigName, r.getStart, r.getEnd, r.getScore)
+    })
+  }
+}
+
+/**
+ * Coverage Recordfor CoverageRDD
```

Comment:
Yeah I feel like there should be an `object Coverage` with methods `private[adam] def apply(feature: Feature)` and `private[adam] def apply(rdd: RDD[Feature])` and then a `def toFeature: Feature` method in the case class.